src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,558 @@
+            //     which suppresses this ID as used/necessary. Also mark the matching restore pragma as used.
+            //  6. Finally, report a diagostic all the pragmas which have not been marked as used.
+
+            var hasPragmaInAnalysisSpan = false;
+            using var _1 = PooledDictionary<string, List<(SyntaxTrivia pragma, bool isDisable)>>.GetInstance(out var idToPragmasMap);
+            using var _2 = ArrayBuilder<(SyntaxTrivia pragma, ImmutableArray<string> ids, bool isDisable)>.GetInstance(out var sortedPragmasWithIds);
+            using var _3 = PooledDictionary<SyntaxTrivia, bool>.GetInstance(out var pragmasToIsUsedMap);
+            using var _4 = ArrayBuilder<string>.GetInstance(out var idsBuilder);
+            using var _5 = PooledHashSet<string>.GetInstance(out var compilerDiagnosticIds);
+            foreach (var trivia in root.DescendantTrivia())
+            {
+                // Check if this is an active pragma with at least one applicable diagnostic ID.
+                if (SyntaxFacts.IsPragmaDirective(trivia, out var isDisable, out var isActive, out var idNodes) &&
+                    isActive &&
+                    idNodes.Count > 0)
+                {
+                    // Iterate through each ID for this pragma and build the supported IDs.
+                    idsBuilder.Clear();
+                    foreach (var idNode in idNodes)
+                    {
+                        // Ignore unsupported IDs and those excluded through user option.
+                        if (!IsSupportedId(idNode, out var id, out var isCompilerDiagnosticId) ||
+                            userExclusions.Contains(id, StringComparer.OrdinalIgnoreCase))
+                        {
+                            continue;
+                        }
+
+                        idsBuilder.Add(id);
+                        if (isCompilerDiagnosticId)
+                        {
+                            compilerDiagnosticIds.Add(id);
+                        }
+
+                        // Add entry to idToPragmasMap
+                        // Insert the pragmas in reverse order for easier processing later.
+                        if (!idToPragmasMap.TryGetValue(id, out var pragmasForIdInReverseOrder))
+                        {
+                            pragmasForIdInReverseOrder = new List<(SyntaxTrivia pragma, bool isDisable)>();
+                            idToPragmasMap.Add(id, pragmasForIdInReverseOrder);
+                        }
+
+                        pragmasForIdInReverseOrder.Insert(0, (trivia, isDisable));
+                    }
+
+                    if (idsBuilder.Count == 0)
+                    {
+                        // No supported ID in this pragma.
+                        continue;
+                    }
+
+                    hasPragmaInAnalysisSpan = hasPragmaInAnalysisSpan || !span.HasValue || span.Value.OverlapsWith(trivia.Span);
```

Comment:
when is `!span.HasValue` true?

Reply:
The span has value when IDE executes analyzers from light bulb context, and is passed as null when IDE executes analyzers from entire file analysis context.